packages/@sanity/code-input/src/CodeInput.tsx

```diff
@@ -86,222 +102,194 @@ function isSupportedLanguage(mode: string) {
-          onBlur={onBlur}
-        />
-      </div>
+      return languageAlternatives.reduce((acc, {title, value: val}) => {
+        const alias = LANGUAGE_ALIASES[val]
+        if (alias) {
+          // eslint-disable-next-line no-console
+          console.warn(
+            `'options.languageAlternatives' lists a language with value "%s", which is an alias of "%s" - please replace the value to read "%s"`,
+            val,
+            alias,
+            alias
+          )
+
+          return acc.concat({title, value: alias})
+        }
+
+        if (!SUPPORTED_LANGUAGES.find((lang) => lang.value === val)) {
+          // eslint-disable-next-line no-console
+          console.warn(
+            `'options.languageAlternatives' lists a language which is not supported: "%s", syntax highlighting will be disabled.`,
+            val
+          )
+        }
+
+        return acc.concat({title, value: val})
+      }, [])
+    }, [type])
+
+    const handleCodeChange = useCallback(
+      (code: string) => {
+        const path = PATH_CODE
+        const fixedLanguage = get(type, 'options.language')
+
+        onChange(
+          PatchEvent.from([
+            setIfMissing({_type: type.name, language: fixedLanguage}),
+            code ? set(code, path) : unset(path),
+          ])
+        )
+      },
+      [onChange, type]
     )
-  }
 
-  handleLanguageFocus = () => {
-    this.props.onFocus(PATH_LANGUAGE)
-  }
-
-  handleCodeFocus = () => {
-    this.props.onFocus(PATH_CODE)
-  }
+    const handleLanguageChange = useCallback(
+      (event: React.ChangeEvent<HTMLSelectElement>) => {
+        const val = event.currentTarget.value
+        const path = PATH_LANGUAGE
 
-  handleFilenameFocus = () => {
-    this.props.onFocus(PATH_FILENAME)
-  }
+        onChange(
+          PatchEvent.from([setIfMissing({_type: type.name}), val ? set(val, path) : unset(path)])
+        )
+      },
+      [onChange, type.name]
+    )
 
-  render() {
-    const {compareValue, value, type, level, readOnly, presence, onBlur, focusPath} = this.props
-    const languages = this.getLanguageAlternatives().slice()
-
-    if (has(type, 'options.language')) {
-      const codeCompareValue = PathUtils.get(compareValue, PATH_CODE)
-      return (
-        <ChangeIndicatorProvider
-          path={PATH_CODE}
-          focusPath={focusPath}
-          value={value?.code}
-          compareValue={codeCompareValue}
-        >
-          <Fieldset legend={type.title} description={type.description} level={level}>
-            {this.renderEditor()}
-          </Fieldset>
-        </ChangeIndicatorProvider>
-      )
-    }
+    const languages = getLanguageAlternatives().slice()
 
     const selectedLanguage =
-      value && value.language ? languages.find((item) => item.value === value.language) : undefined
-
-    if (!selectedLanguage) {
-      languages.unshift({title: 'Select language', value: ''})
-    }
+      props?.value && props?.value.language
```

Comment:
nit: this could be simplified to `props.value?.language` (`props` is always an object)